### DISCUSSION TITLE
JSON-RPC Use

### DISCUSSION THREAD

**Discussion by rdica** [jamulussoftware/jamulus#2691] [Technical Support & Help] [answered]
Hi folks, I'm starting this discussion for those, like myself, that want to play with the JSON-RPC interface in the nightly, and soon to be released, final builds.

https://github.com/jamulussoftware/jamulus/blob/master/docs/JSON-RPC.md

I recall playing with @dtinth branch a while back for a short while, and I thought I was using `curl` but I didn't take notes, so I'm trying again, but keep getting Invalid JSON errors. I can `telnet` to the port and input the same JSON I'm trying to pass with `curl` and that works so I'm likely missing something with the commandline itself:
```
curl -v --http0.9 -d '{"id":1,"method":"jamulus/apiAuth","params":{"secret": "sixteenpluscharsforgrinsngiggles"}}' -H 'Content-Type: application/json' localhost:55124
*   Trying 127.0.0.1:55124...
* Connected to localhost (127.0.0.1) port 55124 (#0)
> POST / HTTP/1.1
> Host: localhost:55124
> User-Agent: curl/7.79.1
> Accept: */*
> Content-Type: application/json
> Content-Length: 80
> 
{"error":{"code":-32700,"message":"Parse error: Invalid JSON received"},"jsonrpc":"2.0"}
* Closing connection 0
```

Now using `telnet`, I tested some modifications of a server, using methods `jamulusserver/setWelcomeMessage` and `jamulusserver/setServerName`. Both methods always return OK per the docs.
Method `jamulusserver/getServerProfile` shows the changes in both the Welcomemessage and Serverame. However, using https://explorer.jamulus.io I don't see the Servername changed, just the Welcomemessage.

1. Any ideas why curl isn't working?
2. Does the Servername change require support from the directory server as well in order for that name to change?

**Comment by pljones** [comment]
1. Missing `}` at the end, I think.  Check with something like `jq`.
2. Currently, the server name is only used when a server registers with the directory.  If you "bounce" the registration (unregister then register), it should pick up the change.  (I can't remember if it's refreshed every time the server pings the directory to confirm it's still alive... it could be...)

**Reply by rdica** [reply]
`jq` was the first thing I used to verify json validity, and the exact same json blob is parsed successfully when I `telnet` to the port instead of using `curl`.

**Comment by hoffie** [comment, marked as answer]
Hi,

> Any ideas why curl isn't working?

The Jamulus JSON-RPC interface is line-based. HTTP won't work (other implementations might use JSON-RPC over HTTP, but Jamulus doesn't). I could imagine that `curl` worked on @dtinth's initial PR by accident, because invalid JSON (e.g. `POST ...`) would just be ignored. The parsing has been made more strict by now though.

> Does the Servername change require support from the directory server as well in order for that name to change?

No additional directory server support is required.

Example:
```
$ nc localhost 8080
{"id":1,"jsonrpc":"2.0","method":"jamulus/apiAuth","params":{"secret": "iyetlE6KgLPUMQ=="}}
{"id":1,"jsonrpc":"2.0","result":"ok"}
```

**Reply by rdica** [reply]
I read up a bit on NDJSON, and tried adding newlines to the data part of the `curl` command, but that didn't seem to work either.
Thanks for the explanations!

**Comment by rdica** [comment]
Interesting, some time after making the servername change around 4:25PM ET (I went to dinner at 5PM, returning about an hour later) and now I dont see my server listed in Any Genre2. Nothing in the server log to indicate unregistering/disconnection. I can connect directly to the server by hostname and IP. There's no firewalling on that port.

**Reply by rdica** [reply]
I noticed that the registrationStatus according to `getServerProfile` was in state `requested`.

I just restarted the server, had a client join it, then changed the welcome message and server name, and I saw the server name update on https://explorer.jamulus.io/ shortly after. registrationStatus shows as `registered`.

A second ServerName change shows registrationStatus as `requested`, and the log file shows
`Server Registration Status update: Registration requested`
and nothing further.

At this point I'm leaving things as they are will continue to monitor both https://explorer.jamulus.io, `getServerProfile`, and serevr log to see if registration completes, and/or the server disappears from explorer again.

**Reply by rdica** [reply]
And the server disappeared from explorer. registrationStatus is still in state `requested`. Nothing further in logs.

**Reply by pljones** [reply]
Have you tried running with the UI and using that to unregister and re-register when this happens?  I'd be interested to note if that fixes it.

Oh and are you on a static IP address?

**Reply by rdica** [reply]
No UI as this is a vps. Built with `qmake-qt5 "CONFIG+=noupcasename headless nosound" Jamulus.pro` on Fedora Server 35.
I looked through the api doc to see if there was a way to manually request registration but didn't find anything.
Yes its a static IP.

Your first question brings a couple things to mind regarding some functionality in the new api.
Will directory server registration/unregistration be added?
While one can alter the welcome message text on the fly, for those that use an html file, will the ability to reread it be added?

**Reply by rdica** [reply]
I'm now testing RPC with 3.9.0beta1. Server is on Any Genre2.
I made three servername changes via RPC and each request completed successfully as monitored by explorer.jamulus.io

In playing with all this I'm simply using `nc localhost 55124` then typing or copy/pasting json lines, and that works fine for this simple testing.
